Jupyterlab 0.31.12

zimbatm opened the pull request:
###### Motivation for this change

updates and ofborg sanity check

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

zimbatm committed:
pythonPackages.jupyterlab_launcher: init at 0.10.5

zimbatm committed:
python3Packages.jupyterlab: 0.4.1 -> 0.31.12

dotlambda commented:
Please rebase on staging: https://groups.google.com/forum/#!topic/nix-devel/IA6o_CRI0NY

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38566&attempt_id=802b0038-4147-4aa8-9129-58d308679b22)

Attempted: python3Packages.jupyterlab

<details><summary>Partial log (click to expand)</summary><p>

```
patching script interpreter paths in /nix/store/8inm9adsvgsj649bjs3drcwnx7dzywc9-python3.6-jupyterlab-0.31.12
/nix/store/gpy3r9ss5ngfkib8ylx7jzgahq7m0x5b-patch-shebangs.sh: line 22: warning: command substitution: ignored null byte in input
/nix/store/gpy3r9ss5ngfkib8ylx7jzgahq7m0x5b-patch-shebangs.sh: line 22: warning: command substitution: ignored null byte in input
/nix/store/gpy3r9ss5ngfkib8ylx7jzgahq7m0x5b-patch-shebangs.sh: line 22: warning: command substitution: ignored null byte in input
/nix/store/gpy3r9ss5ngfkib8ylx7jzgahq7m0x5b-patch-shebangs.sh: line 22: warning: command substitution: ignored null byte in input
checking for references to /tmp/nix-build-python3.6-jupyterlab-0.31.12.drv-0 in /nix/store/8inm9adsvgsj649bjs3drcwnx7dzywc9-python3.6-jupyterlab-0.31.12...
wrapping `/nix/store/8inm9adsvgsj649bjs3drcwnx7dzywc9-python3.6-jupyterlab-0.31.12/bin/jlpm'...
wrapping `/nix/store/8inm9adsvgsj649bjs3drcwnx7dzywc9-python3.6-jupyterlab-0.31.12/bin/jupyter-labhub'...
wrapping `/nix/store/8inm9adsvgsj649bjs3drcwnx7dzywc9-python3.6-jupyterlab-0.31.12/bin/jupyter-lab'...
wrapping `/nix/store/8inm9adsvgsj649bjs3drcwnx7dzywc9-python3.6-jupyterlab-0.31.12/bin/jupyter-labextension'...
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38566&attempt_id=623a16d9-4a26-4272-8da2-15b60a613341)

Attempted: python3Packages.jupyterlab

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/gpy3r9ss5ngfkib8ylx7jzgahq7m0x5b-patch-shebangs.sh: line 22: warning: command substitution: ignored null byte in input
/nix/store/gpy3r9ss5ngfkib8ylx7jzgahq7m0x5b-patch-shebangs.sh: line 22: warning: command substitution: ignored null byte in input
/nix/store/gpy3r9ss5ngfkib8ylx7jzgahq7m0x5b-patch-shebangs.sh: line 22: warning: command substitution: ignored null byte in input
/nix/store/gpy3r9ss5ngfkib8ylx7jzgahq7m0x5b-patch-shebangs.sh: line 22: warning: command substitution: ignored null byte in input
checking for references to /build in /nix/store/w1jg8vyv7yn5hwpyymwl9avn6qawm3xl-python3.6-jupyterlab-0.31.12...
wrapping `/nix/store/w1jg8vyv7yn5hwpyymwl9avn6qawm3xl-python3.6-jupyterlab-0.31.12/bin/jupyter-labhub'...
wrapping `/nix/store/w1jg8vyv7yn5hwpyymwl9avn6qawm3xl-python3.6-jupyterlab-0.31.12/bin/jupyter-labextension'...
wrapping `/nix/store/w1jg8vyv7yn5hwpyymwl9avn6qawm3xl-python3.6-jupyterlab-0.31.12/bin/jupyter-lab'...
wrapping `/nix/store/w1jg8vyv7yn5hwpyymwl9avn6qawm3xl-python3.6-jupyterlab-0.31.12/bin/jlpm'...
/nix/store/w1jg8vyv7yn5hwpyymwl9avn6qawm3xl-python3.6-jupyterlab-0.31.12
```
</p></details>

dotlambda commented on pkgs/development/python-modules/jupyterlab/default.nix:
```diff
@@ -0,0 +1,27 @@
+  ];
+
+  # No tests in archive
+  doCheck = false;
```
Seems like there are tests in `jupyterlab/tests`. However, it seems like node is required:

> Please install nodejs 5+ and npm before continuing.

dotlambda commented on pkgs/development/python-modules/jupyterlab/default.nix:
```diff
@@ -0,0 +1,27 @@
+  meta = with lib; {
+    description = "Jupyter lab environment notebook server extension.";
+    license = with licenses; [ bsd3 ];
+    homepage = "http://jupyter.org/";
```
no quotes

zimbatm replied:
There is no consensus on that in nixpkgs. URL literals get transformed into strings so it doesn't make any practical difference. This is unlike path literals which do have a different type.

FRidh replied:
There is a strong opinion to get rid of the type and use strings instead, but no such (formal) decision has been made. Would it not make sense to keep using the dedicated type for it until decided otherwise? I mean, if people want this they can open an RFC instead of increasing incoherence in the package set.

zimbatm replied:
I agree that we should have the discussion at some point.

nixpkgs is already incoherent so going one way or another doesn't make much difference for new packages. In either case we can write a script that converts existing packages once the above discussion has been resolved.

dotlambda commented on pkgs/development/python-modules/jupyterlab_launcher/default.nix:
```diff
@@ -0,0 +1,25 @@
+  meta = with lib; {
+    description = "This package is used to launch an application built using JupyterLab";
+    license = with licenses; [ bsd3 ];
+    homepage = "http://jupyter.org/";
```
no quotes

GrahamcOfBorg commented:
Failure on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38566&attempt_id=86b53042-b2e2-4c00-86aa-325cdc85ef62)

Attempted: python3Packages.jupyterlab

<details><summary>Partial log (click to expand)</summary><p>

```
cannot build derivation '/nix/store/ggvp6ww093vzhianqwklp745903n5pvq-python3.6-nbformat-4.4.0.drv': 13 dependencies couldn't be built
cannot build derivation '/nix/store/80c22z9cpzl917ld1g442wk3b7z5yq6m-python3.6-pyOpenSSL-17.5.0.drv': 11 dependencies couldn't be built
cannot build derivation '/nix/store/fwzn24m965c9w8xz682hqnlycalvfgvz-python3.6-ipykernel-4.8.2.drv': 13 dependencies couldn't be built
cannot build derivation '/nix/store/8lyzbxr3vbhc7ilkxaggj8cj1y2mnb9s-python3.6-urllib3-1.22.drv': 12 dependencies couldn't be built
cannot build derivation '/nix/store/mg1d3awm5ybk7qmla8i725da3zhkwcj7-python3.6-nbconvert-5.3.1.drv': 22 dependencies couldn't be built
cannot build derivation '/nix/store/9slg9qzmlc4q7l086c61ahpmiqzs8xzn-python3.6-requests-2.18.4.drv': 12 dependencies couldn't be built
cannot build derivation '/nix/store/k8w68s7cji4amywz64qigcnmbiwfz5aw-python3.6-notebook-5.4.1.drv': 22 dependencies couldn't be built
cannot build derivation '/nix/store/h4r8i1iwvyhxasx5r2174j98l65c55f9-python3.6-jupyterlab_launcher-0.10.5.drv': 10 dependencies couldn't be built
cannot build derivation '/nix/store/rn62f6p5zv35qaj1p8yycmbc3bz36y96-python3.6-jupyterlab-0.31.12.drv': 11 dependencies couldn't be built
[31;1merror:[0m build of '/nix/store/rn62f6p5zv35qaj1p8yycmbc3bz36y96-python3.6-jupyterlab-0.31.12.drv' failed
```
</p></details>

zimbatm committed:
Update default.nix

veprbl commented:
I've tested this on macOS. There was a problem with launching ```jupyter-lab```:
```
[I 18:51:45.038 LabApp] JupyterLab beta preview extension loaded from /nix/store/9wck59c4s6jnf1hiln23mw57qlvd4wj2-python3.6-jupyterlab-0.31.12/lib/python3.6/site-packages/jupyterlab
[I 18:51:45.038 LabApp] JupyterLab application directory is /nix/store/mxrl6zan6h63cwl62bd91bgaghz2d94y-python3-3.6.4/share/jupyter/lab
```
It can not determine application directory correctly and that would make it not work. There is also missing (seemingly optional) dependency on nodejs.

Would you mind applying a following fix:
```diff
--- a/pkgs/development/python-modules/jupyterlab/default.nix
+++ b/pkgs/development/python-modules/jupyterlab/default.nix
@@ -1,4 +1,4 @@
-{ lib, buildPythonPackage, isPy3k, fetchPypi, ipython_genutils, jupyterlab_launcher, notebook }:
+{ lib, buildPythonPackage, isPy3k, fetchPypi, ipython_genutils, jupyterlab_launcher, makeWrapper, nodejs, notebook }:
 buildPythonPackage rec {
   pname = "jupyterlab";
   version = "0.31.12";
@@ -9,15 +9,23 @@ buildPythonPackage rec {
     sha256 = "1hp6p9bsr863glildgs2iy1a4l99m7rxj2sy9fmkxp5zhyhqvsrz";
   };

+  nativeBuildInputs = [ makeWrapper ];
   propagatedBuildInputs = [
     ipython_genutils
     jupyterlab_launcher
+    nodejs
     notebook
   ];

   # Depends on npm
   doCheck = false;

+  postInstall = ''
+    for prog in "$out"/bin/*; do
+      wrapProgram "$prog" --set JUPYTERLAB_DIR $out/share/jupyter/lab
+    done
+  '';
+
   meta = with lib; {
     description = "Jupyter lab environment notebook server extension.";
     license = with licenses; [ bsd3 ];
```
Thank you!

veprbl commented:
Other than that it works very nice!

veprbl commented:
@FRidh 
Thank you. I've updated the patch in my original message.

zimbatm committed:
jupyterlab: fix JUPYTERLAB_DIR lookup

without this option jupyterlab will look into the system python installation directories

zimbatm commented:
@veprbl thanks for the wrapper. I didn't add nodejs because it breaks the build in sandbox mode:
```
setting SOURCE_DATE_EPOCH to timestamp 1520372440 of file jupyterlab-0.31.12/PKG-INFO
patching sources
configuring
building
running bdist_wheel
running jsdeps
Installing build dependencies with npm.  This may take a while...
> node /build/jupyterlab-0.31.12/jupyterlab/staging/yarn.js install
yarn install v1.3.2
warning You don't appear to have an internet connection. Try the --offline flag to use the cache for registry queries.
warning Skipping preferred cache folder "/homeless-shelter/.cache/yarn" because it is not writable.
warning Selected the next writable cache folder in the list, will be "/build/.yarn-cache-1000".
```

FRidh commented:
If I am correct `nodejs` is only needed for extensions. We could take that part in a different PR.
